### DISCUSSION TITLE
Expanding the Kafka module

### DISCUSSION THREAD

**Discussion by larsmjoh** [testcontainers/testcontainers-java#4197] [Ideas]
Hi,

I'm currently involved with a client implementing more or less the whole Confluent Kafka stack. We are using Testcontainers to run integration tests locally and that way, have built our own testcontainers for the whole stack (broker, schema registry, rest proxy, kafka connect, ksqldb).
It's an open source-friendly, government client and I believe I can talk them into letting me "give back" the code to the community. 
Would that be of interest?

Regs,
Lars

**Comment by kiview** [comment]
Hey @larsmjoh, thanks for getting in touch. Sorry for not answering earlier, our usage of GH discussions is still new, and therefore I haven't looked in here 😉 

Sharing this knowledge is a great idea and very much appreciated. However, I wonder what is the most valuable way to provide this (with the least maintenance down the road). 
Is there some special code involved in making the different containers work well together in conjunction with Testcontainers, or is it more a kind of convenience wiring? Examples for the former are making changes to the container (or file and config inside the container) during its startup, If the latter is the case, I feel something like a blog post is better suited to document this approach (especially considering that this teaches folks how to set up their own custom stacks in the first place).

Besides this, @bsideup is mostly involved in the existing Kafka module and therefore has a better overview about the specifics.

**Comment by larsmjoh** [comment]
Hi, thanks for getting back.
It's a bit of both, I'd say. The complexity of each container is comparable to the KafkaContainer, with Kafka Connect being the heavier one, since it depends on both Schema Registry and Rest Proxy in addition to the Kafka broker.
We actually coded our own KafkaContainer, while simultaneously learning Kafka, so I think I'm going to rewrite our stack using your container anyway. I'll see what remains after that exercise and if it's worthy of a pull request...
(We also coded an Informix container, but I doubt anyone will be interested in that ;-) )

**Reply by kiview** [reply]
TBH, I think at least documenting what you did with regards to Informix for the community might be extremely valuable. Not only because there might be other users out there (and I think in certain environments, this might be widespread, not an expert about this), but because explaining the pitfalls and workarounds you had to employ is a really great accessible way to enable other users to implement similar stories for their stack.

Again, maybe providing a module is not as useful as a blog post.
